apps/heft/src/plugins/DeleteGlobsPlugin.ts

```diff
@@ -0,0 +1,135 @@
+const HEFT_STAGE_TAP: Tap = {
+  name: PLUGIN_NAME,
+  stage: Number.MIN_SAFE_INTEGER
+} as Tap; /* tappable's typings are wrong here */
+
+export class DeleteGlobsPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    const logger: ScopedLogger = heftSession.requestScopedLogger('delete-globs');
+    heftSession.hooks.clean.tap(PLUGIN_NAME, (clean: ICleanStageContext) => {
+      clean.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+        await this._runDeleteForHeftEvent(
+          HeftEvent.clean,
+          logger,
+          heftConfiguration,
+          clean.properties.pathsToDelete
+        );
+      });
+    });
+
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.preCompile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.compile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.bundle.tap(PLUGIN_NAME, (bundle: IBundleSubstage) => {
+        bundle.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.bundle, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.postBuild.tap(PLUGIN_NAME, (postBuild: IPostBuildSubstage) => {
+        postBuild.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.postBuild, logger, heftConfiguration);
+        });
+      });
+    });
+  }
+
+  private async _runDeleteForHeftEvent(
+    heftEvent: HeftEvent,
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration,
+    additionalPathsToDelete?: Set<string>
+  ): Promise<void> {
+    let deletedFiles: number = 0;
+    let deletedFolders: number = 0;
+
+    const eventActions: IHeftEventActions = await ConfigFile.getConfigConfigFileEventActionsAsync(
+      heftConfiguration
+    );
+
+    const pathsToDelete: Set<string> = new Set<string>(additionalPathsToDelete);
+    for (const deleteGlobsEventAction of eventActions.deleteGlobs.get(heftEvent) || []) {
+      for (const globPattern of deleteGlobsEventAction.globsToDelete) {
+        const resolvedPaths: string[] = await this._resolvePathAsync(
+          globPattern,
+          heftConfiguration.buildFolder
+        );
+        for (const resolvedPath of resolvedPaths) {
+          pathsToDelete.add(resolvedPath);
+        }
+      }
+    }
+
+    await Async.forEachLimitAsync(Array.from(pathsToDelete), 100, async (pathToDelete) => {
```

Comment:
Should this constant be stored in a central place and so it can be used for all such parallel tasks?  And (later) maybe it should be configurable?

For example: [#1934 - [rush] "rush update" exceeds "maxfiles" limit on macOS ](https://github.com/microsoft/rushstack/issues/1934)

Reply:
Sure - we can do that. This isn't new code (this is moved code), so we should do that separately.